packages/jupyter-ai/src/inline-completions.ts

```diff
@@ -0,0 +1,500 @@
+        },
+        maxSuffix: {
+          title: 'Maximum suffix length',
+          minimum: 0,
+          type: 'number',
+          description:
+            'At most how many suffix characters should be provided to the model.'
+        },
+        disabledLanguages: {
+          title: 'Disabled languages',
+          type: 'array',
+          items: {
+            type: 'string',
+            oneOf: knownLanguages.map(language => {
+              return { const: language.name, title: displayName(language) };
+            })
+          },
+          description:
+            'Languages for which the completions should not be shown.'
+        }
+      },
+      default: JupyterAIInlineProvider.DEFAULT_SETTINGS as any
+    };
+  }
+
+  async configure(settings: { [property: string]: JSONValue }): Promise<void> {
+    this._settings = settings as unknown as JupyterAIInlineProvider.ISettings;
+  }
+
+  isEnabled(): boolean {
+    return this._settings.enabled;
+  }
+
+  isLanguageEnabled(language: string) {
+    return !this._settings.disabledLanguages.includes(language);
+  }
+
+  /**
+   * Extract prefix from request, accounting for context window limit.
+   */
+  private _prefixFromRequest(request: CompletionHandler.IRequest): string {
+    const textBefore = request.text.slice(0, request.offset);
+    const prefix = textBefore.slice(
+      -Math.min(this._settings.maxPrefix, textBefore.length)
+    );
+    return prefix;
+  }
+
+  /**
+   * Extract suffix from request, accounting for context window limit.
+   */
+  private _suffixFromRequest(request: CompletionHandler.IRequest): string {
+    const textAfter = request.text.slice(request.offset);
+    const prefix = textAfter.slice(
+      0,
+      Math.min(this._settings.maxPrefix, textAfter.length)
+    );
+    return prefix;
+  }
+
+  private _resolveLanguage(language: IEditorLanguage | null) {
+    if (!language) {
+      return 'plain English';
+    }
+    if (language.name === 'ipythongfm') {
+      return 'markdown';
+    }
+    return language.name;
+  }
+
+  private _settings: JupyterAIInlineProvider.ISettings =
+    JupyterAIInlineProvider.DEFAULT_SETTINGS;
+
+  private _currentModel = '';
+  private _counter = 0;
+}
+
+namespace JupyterAIInlineProvider {
+  export interface IOptions {
+    completionHandler: CompletionWebsocketHandler;
+    languageRegistry: IEditorLanguageRegistry;
+  }
+  export interface ISettings {
+    maxPrefix: number;
+    maxSuffix: number;
+    debouncerDelay: number;
+    enabled: boolean;
+    disabledLanguages: string[];
+  }
+  export const DEFAULT_SETTINGS: ISettings = {
+    maxPrefix: 10000,
+    maxSuffix: 10000,
+    // The debouncer delay handling is implemented upstream in JupyterLab;
+    // here we just increase the default from 0, as compared to kernel history
+    // the external AI models may have a token cost associated.
+    debouncerDelay: 250,
+    enabled: true,
```

Comment:
I am not 100% sure if we want to make it opt-in or opt-out but I _think_ it should be opt-out at least in the first release as we iron any rough edges. Do you agree?

```suggestion
    enabled: false,
```

Reply:
Hm, I think that we will definitely want to add this control in JAI's Settings (i.e. the UI where you set the language & embedding models in chat). I can help with the implementation of such controls later.

As long as both 1) JupyterLab v4.1.0 has an official release (so we don't depend on an alpha version), and 2) users can easily toggle this in JAI's Settings, then I think we can have this feature enabled by default in its very first release. Why not let users enjoy it faster? 😁 ❤️

Reply:
I would strongly prefer to keep these settings in global settings where all completion settings are (rather than duplicating code to add this in chat UI settings to). We could add a button in the chat UI settings which would open Settings Editor on the inline completer tab, what do you think?